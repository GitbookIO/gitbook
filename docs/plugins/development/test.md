# Testing your plugin

### On local machine

Testing your plugin on your book before publishing is possible using [npm link](https://docs.npmjs.com/cli/link).

In the plugin's folder, run:

```
$ npm link
```

then in your book's folder:

```
$ npm link gitbook-plugin-<plugin's name>
```

### Unit testing on Travis

[gitbook-tester](https://github.com/todvora/gitbook-tester) makes it easy to write **Node.js/Mocha** unit tests for your plugins. Using [Travis.org](https://travis.org), tests can be run on each commit/tag.
