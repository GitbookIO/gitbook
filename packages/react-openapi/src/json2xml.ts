import { jsXml } from 'json-xml-parse';

/**
 * This function converts an object to XML.
 */
export function json2xml(data: Record<string, any>) {
    return jsXml.toXmlString(data, { beautify: true });
}
