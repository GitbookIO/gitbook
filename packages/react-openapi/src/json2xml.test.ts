import { describe, expect, it } from 'bun:test';

import { json2xml } from './json2xml';

describe('getUrlFromServerState', () => {
    it('transforms JSON to xml', () => {
        const xml = json2xml({
            foo: 'bar',
        });

        expect(xml).toBe('<?xml version="1.0"?>\n<foo>bar</foo>\n');
    });

    it('wraps array items', () => {
        const xml = json2xml({
            urls: {
                url: ['https://example.com', 'https://example.com'],
            },
        });

        expect(xml).toBe(
            '<?xml version="1.0"?>\n<urls>\n\t<url>https://example.com</url>\n\t<url>https://example.com</url>\n</urls>\n'
        );
    });

    it('indents correctly', () => {
        const xml = json2xml({
            id: 10,
            name: 'doggie',
            category: {
                id: 1,
                name: 'Dogs',
            },
            photoUrls: ['string'],
            tags: [
                {
                    id: 0,
                    name: 'string',
                },
            ],
            status: 'available',
        });

        expect(xml).toMatchSnapshot();
    });
});
