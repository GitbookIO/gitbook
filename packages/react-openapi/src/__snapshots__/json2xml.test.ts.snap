// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getUrlFromServerState indents correctly 1`] = `
"<?xml version="1.0"?>
<id>10</id>
<name>doggie</name>
<category>
	<id>1</id>
	<name>Dogs</name>
</category>
<photoUrls>string</photoUrls>
<tags>
	<id>0</id>
	<name>string</name>
</tags>
<status>available</status>
"
`;
